### PR TITLE
feat(model): #36 one dimensional harmonic oscillator chain

### PR DESCRIPTION
**Pull request by cmp0xff**
with periodic boundary condition
- [x] Dependency: #72

### CODE REVIEW

File: hamilflow/models/d0/complex_harmonic_oscillator.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+from dataclasses import asdict, dataclass, field
+from functools import cached_property
+from typing import Mapping
+
+import numpy as np
+from numpy.typing import ArrayLike
+from pydantic import BaseModel, Field, computed_field
+
+
+class ComplexHarmonicOscillatorSystem(BaseModel):
```
Comment (GrumpySapiens) [inline hamilflow/models/d0/complex_harmonic_oscillator.py:10]: Out of curiosity, why not reuse the HarmonicOscillatorSystem class? Seems like you've extend the HarmonicOscillatorSystem to support complex cases.

File: hamilflow/models/d1/harmonic_oscillator_chain.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    ) -> None:
+        self.omega = omega
+        self.dof = len(half_initial_conditions) + odd_dof
+        self.travelling_waves = [
+            SimpleComplexHarmonicOscillator(
+                dict(
+                    omega=2 * omega * np.sin(np.pi * k / self.dof),
+                    real=not odd_dof and k == len(half_initial_conditions),
+                ),
+                dict(x0=ic["y0"], phi=ic["phi"]),
```
Comment (GrumpySapiens) [inline hamilflow/models/d1/harmonic_oscillator_chain.py:57]: Wondering if you need to enforce the initial condition to use `y0` for amplitude. (I mean some people might be using y, like me...)

File: hamilflow/models/d0/complex_harmonic_oscillator.py (hunk 1 above)
Comment (cmp0xff) [reply hamilflow/models/d0/complex_harmonic_oscillator.py]: https://github.com/kausalflow/hamilflow/pull/58/commits/bc73c3320a2a8c89aae3780cb2ddafb4a305a992

File: hamilflow/models/d1/harmonic_oscillator_chain.py (hunk 2 above)
Comment (cmp0xff) [reply hamilflow/models/d1/harmonic_oscillator_chain.py]: https://github.com/kausalflow/hamilflow/pull/58/commits/bc73c3320a2a8c89aae3780cb2ddafb4a305a992 now I use `amp`

File: hamilflow/models/d1/harmonic_oscillator_chain.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    Since the original degrees of freedom are real, the initial conditions of the propagating waves need to satisfy
+    $Y_k = Y^*_{-k \mod N}$, see [Wikipedia](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#DFT_of_real_and_purely_imaginary_signals).
+    """
+
+    def __init__(
+        self,
+        omega: float,
+        half_initial_conditions: Collection[Mapping[str, complex | float | int]],
+        odd_dof: bool,
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (cmp0xff) [reply hamilflow/models/d1/harmonic_oscillator_chain.py]: https://github.com/kausalflow/hamilflow/pull/58/commits/7ceefc8d86c97657b08c88932ea121f71ebd682a

File: mkdocs.yml [hunk 4]
```diff
@@ -84,10 +84,12 @@ nav:
...
     - "Brownian Motion": tutorials/brownian_motion.py
     - "Pendulum": tutorials/pendulum.py
+    - "Harmonic Oscillator Chain": tutorials/harmonic_oscillator_chain.py
   - References:
     - "Introduction": references/index.md
     - "Models":
       - "Harmonic Oscillator": references/models/harmonic_oscillator.md
       - "Brownian Motion": references/models/brownian_motion.md
       - "Pendulum": references/models/pendulum.md
+      - "Harmonic Oscillator Chain": references/models/harmonic_oscillator_chain.py
```
Comment (emptymalei) [inline mkdocs.yml:94]:
```suggestion
      - "Harmonic Oscillator Chain": references/models/harmonic_oscillator_chain.md
```

File: hamilflow/models/harmonic_oscillator_chain.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+from .free_particle import FreeParticle
+from .harmonic_oscillator import ComplexSimpleHarmonicOscillator
+
+
+class HarmonicOscillatorsChain:
+    r"""Generate time series data for a coupled harmonic oscillator chain
+    with periodic boundary condition.
+
+    A one-dimensional circle of $N$ interacting harmonic oscillators can be described by the Lagrangian action
+    $$S_L[x_i] = \int_{t_0}^{t_1}\mathbb{d} t \left\\{ \sum_{i=0}^{N-1} \frac{1}{2}m \dot x_i^2 - \frac{1}{2}m\omega^2\left(x_i - x_{i+1}\right)^2 \right\\}\\,,$$
```
Comment (emptymalei) [inline hamilflow/models/harmonic_oscillator_chain.py:18]:
```suggestion
    $$S_L[x_i] = \int_{t_0}^{t_1}\mathbb{d} t \left\{ \sum_{i=0}^{N-1} \frac{1}{2}m \dot x_i^2 - \frac{1}{2}m\omega^2\left(x_i - x_{i+1}\right)^2 \right\}\,,$$
```

File: hamilflow/models/harmonic_oscillator_chain.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    $Y_k = Y^*_{-k \mod N}$, see [Wikipedia](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#DFT_of_real_and_purely_imaginary_signals).
+    """
+
+    def __init__(
+        self,
+        omega: float,
+        initial_conditions: Sequence[Mapping[str, float | tuple[float, float]]],
+        odd_dof: bool,
+    ) -> None:
+        """Instantiate an oscillator chain.
```
Comment (emptymalei) [inline hamilflow/models/harmonic_oscillator_chain.py:33]: Minor issues but is is easier to read and use if we move the docstring for init into the  docstring of the class.

File: hamilflow/models/harmonic_oscillator_chain.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    with periodic boundary condition.
+
+    A one-dimensional circle of $N$ interacting harmonic oscillators can be described by the Lagrangian action
+    $$S_L[x_i] = \int_{t_0}^{t_1}\mathbb{d} t \left\\{ \sum_{i=0}^{N-1} \frac{1}{2}m \dot x_i^2 - \frac{1}{2}m\omega^2\left(x_i - x_{i+1}\right)^2 \right\\}\\,,$$
+    where $x_N \coloneqq x_0$.
+
+    This system can be solved in terms of _travelling waves_, obtained by discrete Fourier transform.
+
+    Since the original degrees of freedom are real, the initial conditions of the propagating waves need to satisfy
+    $Y_k = Y^*_{-k \mod N}$, see [Wikipedia](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#DFT_of_real_and_purely_imaginary_signals).
```
Comment (emptymalei) [inline hamilflow/models/harmonic_oscillator_chain.py:24]: I think it is a bit easier for our future selves if we have have a usage example in the docstring.
Something like

https://github.com/kausalflow/hamilflow/blob/04ea13c523f8beb7d48123754ec96251b98b5c49/hamilflow/models/harmonic_oscillator.py#L141-L151

File: hamilflow/models/harmonic_oscillator_chain.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            omega=self.omega,
+            n_dof=self.n_dof,
+            free_mode=self.free_mode.definition,
+            independent_csho_modes=[
+                rwm.definition for rwm in self.independent_csho_modes
+            ],
+        )
+
+    def _z(
+        self, t: "Sequence[float] | ArrayLike[float]"
```
Comment (emptymalei) [inline hamilflow/models/harmonic_oscillator_chain.py:98]: I just noticed this. Is there a reason that we use quotes here? Was there a parsing error?

Comment (cmp0xff) [reply hamilflow/models/harmonic_oscillator_chain.py:98]: py310 can easily get unhappy without the quotation marks for no obvious reason. Is there an argument against using quotation marks?

Comment (emptymalei) [reply hamilflow/models/harmonic_oscillator_chain.py:98]: > py310 can easily get unhappy without the quotation marks for no obvious reason. Is there an argument against using quotation marks?

not sure, just thinking that the hint becomes a string and not sure if mypy or IDE works in the way it intended to.

I know people annotate returned self using str but not sure what happens for args.

**Comment by emptymalei** [comment]
So, I ran the tutorial, looks reasonable from an intuition perspective. I played with different initial conditions and they all seem to make sense.

File: hamilflow/models/harmonic_oscillator_chain.py (hunk 8 above)
Comment (cmp0xff) [reply hamilflow/models/harmonic_oscillator_chain.py:98]: > > py310 can easily get unhappy without the quotation marks for no obvious reason. Is there an argument against using quotation marks?
> 
> not sure, just thinking that the hint becomes a string and not sure if mypy or IDE works in the way it intended to.
> 
> I know people annotate returned self using str but not sure what happens for args. 

- vscode works fine with the hint being a string. 
- it seems quoted hints belong to so-called forward reference https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Comment (emptymalei) [reply hamilflow/models/harmonic_oscillator_chain.py:98]: Then it is good.

File: hamilflow/models/harmonic_oscillator_chain.py (hunk 7 above)
Comment (cmp0xff) [reply hamilflow/models/harmonic_oscillator_chain.py:24]: Issue created https://github.com/kausalflow/hamilflow/issues/73

**Review by emptymalei — APPROVED** [review]

File: hamilflow/models/harmonic_oscillator_chain.py (hunk 6 above)
Comment (cmp0xff) [reply hamilflow/models/harmonic_oscillator_chain.py]: https://github.com/kausalflow/hamilflow/pull/58/commits/84462b0a1860398f2d16900549f8ae64aeff60a6

**Pull request merged**